File: 2349. Design a Number Container System/Solution.py
import heapq
from collections import defaultdict


class NumberContainers:

    def __init__(self):
        self.data = {}  # Maps index to number
        self.num2idx = defaultdict(list)

    def change(self, index: int, number: int) -> None:
        self.data[index] = number
        heapq.heappush(self.num2idx[number], index)

    def find(self, number: int) -> int:
        if number not in self.num2idx:
            return -1
        while self.num2idx[number] and self.data[self.num2idx[number][0]] != number:
            heapq.heappop(self.num2idx[number])
        if not self.num2idx[number]:
            self.num2idx.pop(number, None)
            return -1
        return self.num2idx[number][0]


# Your NumberContainers object will be instantiated and called as such:
# obj = NumberContainers()
# obj.change(index,number)
# param_2 = obj.find(number)

File: 2349. Design a Number Container System/Solution.cpp
#include <set>
#include <unordered_map>
#include <vector>
using namespace std;

class NumberContainers {
public:
    NumberContainers()
    {
    }

    void change(int index, int number)
    {
        if (!data.contains(index)) {
            data[index] = number;
            num2idx[number].emplace(index);
        } else if (number != data[index]) {
            int old_val = data[index];
            num2idx[old_val].erase(index);
            if (num2idx[old_val].empty())
                num2idx.erase(old_val);
            data[index] = number;
            num2idx[number].emplace(index);
        }
    }

    int find(int number)
    {
        if (!num2idx.contains(number))
            return -1;
        return *num2idx[number].begin();
    }

private:
    unordered_map<int, int> data;
    unordered_map<int, set<int>> num2idx;
};

/**
 * Your NumberContainers object will be instantiated and called as such:
 * NumberContainers* obj = new NumberContainers();
 * obj->change(index,number);
 * int param_2 = obj->find(number);
 */
